### DISCUSSION TITLE
Print complete & change filament (M600) notification

### DISCUSSION THREAD

**Discussion by blonestar** [nlef/moonraker-telegram-bot#201] [Q&A] [answered]
Can someone point me where to read how to setup to get notification on print complete and for example on change filament with gcode (macro M600).
Thank you

```
# I tried this for print complete
[progress_notification]
percent: 100
```

```
######################################################################
# Filament Change
######################################################################

# M600: Filament Change. This macro will pause the printer, move the
# tool to the change position, and retract the filament 50mm. Adjust
# the retraction settings for your own extruder. After filament has
# been changed, the print can be resumed from its previous position
# with the "RESUME" gcode.

[pause_resume]

[gcode_macro M600]
gcode:
    {% set X = params.X|default(50)|float %}
    {% set Y = params.Y|default(0)|float %}
    {% set Z = params.Z|default(10)|float %}
    SAVE_GCODE_STATE NAME=M600_state
    PAUSE
    G91
    G1 E-.8 F2700
    G1 Z{Z}
    G90
    G1 X{X} Y{Y} F3000
    G91
    G1 E-50 F1000
    RESTORE_GCODE_STATE NAME=M600_state
```

**Comment by blonestar** [comment]
This doesn't work for me either. Command has been executed without error but no message in Telegram.
![image](https://user-images.githubusercontent.com/5240133/191616874-9742c23e-11f7-4e26-95a0-39388f5d148d.png)

**Reply by blonestar** [reply]
Bot is configured and working, I would say normally. Secrets are moved to secrets.conf
So that is not an issue, and I need a hint. :)

![image](https://user-images.githubusercontent.com/5240133/191618402-8d5e20fd-d043-4683-8bc5-e77dbafda145.png)

**Comment by nlef** [comment]
> This doesn't work for me either. Command has been executed without error but no message in Telegram.
> ![image](https://user-images.githubusercontent.com/5240133/191616874-9742c23e-11f7-4e26-95a0-39388f5d148d.png)
> 

Extra symbol ':' must be removed. Prefix must be 'telegram'.

**Reply by blonestar** [reply]
Thanks but it's not working, just tested. Restarted everything, command sent but Telegram is quiet.
![image](https://user-images.githubusercontent.com/5240133/191726849-31f721fd-7ab9-4355-ab13-23b7eadd8656.png)
![image](https://user-images.githubusercontent.com/5240133/191728582-cb107659-cda1-44d5-bf55-315045de6ffd.png)
![image](https://user-images.githubusercontent.com/5240133/191729135-dabca26f-676e-4ae4-bf1a-ccb0cb2f89f7.png)

**Comment by blonestar** [comment, marked as answer]
Oh, I figure out...
It should be 'tgnotify' instead of 'telegram'.

`RESPOND PREFIX=tgnotify MSG="Test message"`

**Comment by blonestar** [comment]
Still not sure how to send message on printing job complete.

**Reply by aka13-404** [reply]
How do you do anything when your print is complete? You add a specific gcode or instruction to your "end gcode" in your slicer...

**Reply by blonestar** [reply]
I'm new to 3d printing, gcode and Klipper. Forgive me. :)
But understand now, will probably add end gcode to my Klipper printer config, thank you very much.

**Reply by aka13-404** [reply]
If it helps, the bot is rather an executing organ in this whole ecosystem. It rarely does stuff on its own, on purpose. The general idea is that you instruct it to do something, when you need it. This "when" is rarely easily detectable without some external variable, or withour straight injection from the gcode being printed.

**Reply by blonestar** [reply]
Yep, getting the picture day by day. :)
And btw it works flawlessly, thank you for the great work!
![image](https://user-images.githubusercontent.com/5240133/191775785-4b0b0aee-8a90-46cf-b8ae-5cc5c9b072cf.png)
